js_modules/dagster-ui/packages/ui-core/src/runs/useCursorAccumulatedQuery.tsx

```diff
@@ -1,45 +1,125 @@
-import {useQuery} from '@apollo/client';
+import {OperationVariables, useApolloClient} from '@apollo/client';
 import {DocumentNode} from 'graphql';
-import {useEffect, useState} from 'react';
+import {useEffect, useMemo, useState} from 'react';
 
-import {useQueryRefreshAtInterval} from '../app/QueryRefresh';
+import {useRefreshAtInterval} from '../app/QueryRefresh';
 
-type FetchState = {
+type FetchResult<DataType, CursorType, ErrorType> = {
+  data: DataType[];
   hasMore: boolean;
-  cursor: string | null | undefined;
+  cursor: CursorType | undefined;
+  error: ErrorType | undefined;
 };
 
-export function useCursorAccumulatedQuery<T, TVars extends {cursor?: string | null}, U>(options: {
+type FetcherFunction<DataType, CursorType, ErrorType> = (
+  cursor: CursorType | undefined,
+) => Promise<FetchResult<DataType, CursorType, ErrorType>>;
+
+class AccumulatingDataFetcher<DataType, CursorType, ErrorType> {
+  private fetchData: FetcherFunction<DataType, CursorType, ErrorType>;
+  private onData: (data: DataType[]) => void;
+  private onError?: (error: ErrorType) => void;
+
+  private hasMoreData = true;
+  private dataSoFar: DataType[] = [];
+  private currentCursor: CursorType | undefined = undefined;
```

Comment:
same

Reply:
Ahh so these need to be instance vars so that calling `fetch` multiple times picks up where it left off rather than restarting from the beginning. We want it the periodic calls to `fetch` to make just the last request, pulling in any new data and advancing the cursor.

Reply:
Ahhh makes sense!